labrad/protocol.py

```diff
@@ -333,5 +373,92 @@ def __init__(self, source, context, target):
     def __repr__(self):
         return 'MessageContext(source=%s, ID=%s, target=%s)' % (self.source, self.ID, self.target)
 
-# factory for churning out LabRAD connections
+
+# factory for creating LabRAD connections
 factory = protocol.ClientCreator(reactor, LabradProtocol)
+
+
+@inlineCallbacks
+def connect(host=C.MANAGER_HOST, port=None, tls=C.MANAGER_TLS):
+    """Connect to LabRAD and return a deferred that fires the protocol object.
+
+    Args:
+        host (str): The hostname of the manager.
+        port (int): The tcp port of the manager. If None, use the appropriate
+            default value based on the TLS mode.
+        tls (str): The tls mode to use for this connection. See:
+            `labrad.constants.check_tls_mode`.
+
+    Returns:
+        twisted.internet.defer.Deferred(LabradProtocol): A deferred that will
+        fire with the protocol once the connection is established.
+    """
+    tls = C.check_tls_mode(tls)
+    if port is None:
+        port = C.MANAGER_PORT_TLS if tls == 'on' else C.MANAGER_PORT
+
+    if tls == 'on':
+        tls_options = crypto.tls_options(host)
+        p = yield factory.connectSSL(host, port, tls_options, timeout=C.TIMEOUT)
+        returnValue(p)
+
+    def do_connect():
+        return factory.connectTCP(host, port, timeout=C.TIMEOUT)
+
+    @inlineCallbacks
+    def start_tls(p, cert_string=None):
+        try:
+            resp = yield p.sendRequest(C.MANAGER_ID, [(1L, ('STARTTLS', host))])
+        except Exception:
+            raise Exception(
+                'Failed sending STARTTLS command to server. You should update '
+                'the manager and configure it to support encryption or else '
+                'disable encryption for clients. See '
+                'https://github.com/labrad/pylabrad/blob/master/CONFIG.md')
+        cert = resp[0][1]
+        p.transport.startTLS(crypto.tls_options(host, cert_string=cert_string))
+        returnValue(cert)
+
+    def ping(p):
+        return p.sendRequest(C.MANAGER_ID, [(2L, 'PING')])
+
+    p = yield do_connect()
+    is_local_connection = util.is_local_connection(p.transport)
+    if ((tls == 'starttls-force') or
+        (tls == 'starttls' and not is_local_connection)):
+        try:
+            cert = yield start_tls(p)
+        except Exception:
+            # TODO: remove this retry. This is a temporary fix to support
+            # compatibility until TLS is fully deployed.
+            print ('STARTTLS failed; will retry without encryption in case we '
+                   'are connecting to a legacy manager.')
+            p = yield connect(host, port, tls='off')
+            print 'Connected without encryption.'
+            returnValue(p)
+        try:
+            yield ping(p)
```

Comment:
This is a hack around calling `startTLS` on a twisted transport returns nothing. Ideally it would return a deferred whose callback or errback would trigger when the TLS negotiation either succeeds or fails. Instead, we have to actively try to do something with the channel (e.g. send a ping message) in order to find out what happened.